astropy/config/paths.py

```diff
@@ -167,20 +185,34 @@ def _get_dir_path(cls, rootname: str) -> Path:
                 path.mkdir(exist_ok=True)
             return path.resolve()
 
-        if (
-            (dir_ := os.getenv(cls._directory_env_var)) is not None
-            and (xch := Path(dir_)).exists()
-            and not (xchpth := xch / rootname).is_symlink()
-        ):
-            if xchpth.exists():
-                return xchpth.resolve()
+        if (env_dir_str := os.getenv(cls._directory_env_var)) is None:
+            return cls._find_or_create_root_dir(linkto=None, pkgname=rootname)
 
-            # symlink will be set to this if the directory is created
-            linkto = xchpth
-        else:
+        warn = partial(
+            cls._warn_env_var_is_ignored, env_var_name=cls._directory_env_var
+        )
+        if not (env_dir_path := Path(env_dir_str)).is_dir():
+            warn("no such directory was found")
+            return cls._find_or_create_root_dir(linkto=None, pkgname=rootname)
+
+        path = env_dir_path / rootname
+        if path.is_symlink():
+            return cls._find_or_create_root_dir(linkto=None, pkgname=rootname)
+        elif path.is_dir():
+            return path.resolve()
+
+        linkto: Path | None
```

Comment:
Is this for typing? Why is there type declaration in the middle of the function, especially when linkto is already used above?